File: websockets/websockets_emscripten.cpp
#include "websockets.h"

namespace sgf {

// ***** WebSocket constructor *****

EM_CALLBACK void sgfWSOnOpen(WebSocket* ws_ptr) {
	ws_ptr->open.emit();
}

EM_CALLBACK void sgfWSOnClose(WebSocket* ws_ptr) {
	ws_ptr->closed.emit();
}

EM_CALLBACK void sgfWSOnError(WebSocket* ws_ptr, const char* err_cstr) {
	ws_ptr->error.emit(String(err_cstr));
}

EM_CALLBACK void sgfWSOnMessage(WebSocket* ws_ptr, const char* msg_cstr) {
	ws_ptr->message.emit(String(msg_cstr));
}

EM_EXTERN void sgfWSCreate(WebSocket* ws_ptr, const char* url_cstr);

WebSocket::WebSocket(CString url) { //
	sgfWSCreate(this, url.c_str());
}

// ***** WebSocket destructor *****

EM_EXTERN void sgfWSDestroy(WebSocket* ws_ptr);

WebSocket::~WebSocket() { //
	sgfWSDestroy(this);
}

// ***** WebSocket::send *****

EM_EXTERN void sgfWSSend(WebSocket* ws_ptr, const char* msg_cstr);

void WebSocket::send(CString msg) { //
	sgfWSSend(this, msg.c_str());
}

// ***** WebSocket::close *****

EM_EXTERN void sgfWSClose(WebSocket* ws_ptr);

void WebSocket::close() { //
	sgfWSClose(this);
}

} // namespace sgf

File: websockets/websockets_emscripten.js
mergeInto(LibraryManager.library, {

    sgfWSCreate: function (ws_ptr, url_cstr) {

        let ws;

        try {
            ws = new WebSocket(UTF8ToString(url_cstr));
            _sgfRegisterObject(ws, ws_ptr);
        } catch (error) {
            console.log("Exception creating WebSocket", error);
            return;
        }

        ws.addEventListener('open', (event) => {
            _sgfWSOnOpen(ws_ptr);
        });

        ws.addEventListener('close', (event) => {
            _sgfWSOnClose(ws_ptr);
        });

        ws.addEventListener('error', (event) => {
            const error = JSON.stringify(event, Object.getOwnPropertyNames(event));
            const error_cstr = allocateUTF8(error);
            _sgfWSOnError(ws_ptr, error_cstr);
            _free(error_cstr);
        });

        ws.addEventListener('message', (event) => {
            const msg_cstr = allocateUTF8(event.data);
            _sgfWSOnMessage(ws_ptr, msg_cstr);
            _free(msg_cstr);
        });
    },

    sgfWSDestroy: function (ws_ptr) {

        const ws = _sgfGetObject(ws_ptr);
        _sgfDeregisterObject(ws_ptr);

        ws.close();
    },

    sgfWSSend: function (ws_ptr, msg_cstr) {

        const ws = _sgfGetObject(ws_ptr);

        ws.send(UTF8ToString(msg_cstr));
    },

    sgwWSClose: function (ws_ptr) {

        const ws = _sgfGetObject(ws_ptr);

        ws.close();
    }

});
